roll/ast.py

```diff
@@ -0,0 +1,685 @@
+import logging
+import random
+from enum import Enum, auto
+from functools import wraps
+
+import roll.exceptions as rollerr
+from roll.types import *
+from utils.exceptions import OutputTooLargeError
+
+MAX_ROLLS = 1000
+
+
+def isfunction(token):
+    return isinstance(token, TokenFunction)
+
+
+class Environment:
+    def __init__(self, root):
+        self.root = root
+        self.closure = {}
+        self.trace = []
+
+    def copy(self):
+        out = Environment(self.root)
+        out.closure = self.closure.copy()
+        out.trace = self.trace
+        return out
+
+
+class HashCounter:
+    """A mutating object that records the variable IDs to be used in general hashing and eta-reduction hashing"""
+
+    def __init__(self):
+        self.__next_id = 0
+        self.__next_scope_id = -1
+        self.trace = []
+
+    def copy(self):
+        out = HashCounter()
+        out.__next_id = self.__next_id
+        out.__next_scope_id = self.__next_scope_id
+        out.trace = self.trace
+        return out
+
+    @property
+    def next_id(self):
+        return self.__next_id
+
+    def pop_id(self):
+        id = self.__next_id
+        self.__next_id += 1
+        return id
+
+    @property
+    def next_scope_id(self):
+        return self.__next_scope_id
+
+    def pop_scope_id(self):
+        id = self.__next_scope_id
+        self.__next_scope_id -= 1
+        return id
+
+
+def trace(func):
+    """A decorator that updates the environment/counter to know its position in the current expression"""
+
+    def wrapper(*args, **kwargs):
+        # args[0] will be self (the expression token object)
+        # args[1] should always be the env/map object
+        args[1].trace.append(args[0])
+        out = func(*args, **kwargs)
+        args[1].trace.pop()
+        return out
+
+    return wrapper
+
+
+class Token:

```

Comment:
This appears to be an abstract base class - inherit from https://docs.python.org/3/library/abc.html#abc.ABC